IBM/FHIR
# Bulk Data Enhancement

- [x] [BulkData Import Operation: Switch S3 and File to use Byte Specific
read locations #2549](https://github.com/IBM/FHIR/issues/2549)
	- *FILE* Provider
		- Convert to use SeekableByteChannel with FileProvider
		- Enables skipping to a byte versus read and discard buffer approach
		- Use an upper bound read BLOCK to 512K
		- Improved Read Performance from 33 Resources per second to 66 per
second in local tests
	- Increase Import READ from a block of 20 to 50 (matrixes with block
size)

- [x] Import ChunkReader changed name of wrapper to provider
- [x] [ImportPartitionAnalyzer not reporting infly rate on threshold
#3000](https://github.com/IBM/FHIR/issues/3000)
	- ImportPartitionAnalyzer - modify the infly rate reporting message and
changed the message to report out Resources/second and report which
workitem is triggering the message.
		- Improve accuracy of the infly rate calculation
		- Sample Message: [INFO    ] Import in-fly rate:
[286/286/r4_AllergyIntolerance.ndjson/AllergyIntolerance]
reportingThreshold=[2000] resources imported in 12610 milliseconds,
ImportRate: [0.064] Resources/milliseconds
			- Instance/Execution/File/ResourceType threshold milliseconds and
rate.
- [x] [BulkData Import Operation: Switch S3 and File to use Byte Specific
read locations #2549](https://github.com/IBM/FHIR/issues/2549)
	- *S3* Provider
		- Switch to retrieve bytes by range
		- Added tests to test CountInputStreamReader to test
			- Blocks of 32K buffered while using readLine
			- Multiple Lines
			- Stripping CR from the Input
		- Establish Drain process for S3ObjectInputStream before closing it
- [x] [Failed to execute the s3 access, check the s3 configuration is a bit
misleading #3003](https://github.com/IBM/FHIR/issues/3003)
	- Relaxed the S3Preflight to support http connection with warning
		- MSG: BulkData Request is using a provider that connects to an
insecure url: http://bucket.bucket:9000
	- Improved the Response Handling to capture the ex and package it back
up the stack
- [x] [Bulk data import with missing "id"
#3016](https://github.com/IBM/FHIR/issues/3016)
	- Switched the conditionalUpdate to skip and pass in the oldResource as
they exactly match rather than the Meta updated version with the unused
new version
	- Update the Create to ignore the Version when passed in
	- Enable the Version on Update by using the Create with Version.
	- Refactor the BulkAuditLogger to eliminate the dual resource usage
	- Updated the documentation to explain the before on $import
- [x] POM Change: Exclude duplicate dependency with multiple SLF4J bindings on (excluded
log4j logger) -- Reverted this as it caused a stackoverflow error. Add comments to the pom.xml 
- [x] [BulkData Import Operation: Switch S3 and File to use Byte Specific
read locations #2549](https://github.com/IBM/FHIR/issues/2549)
	- *S3* Provider
		- Implement Range based retrieval in the S3Provider
		- Added a Test Provider Main to suck content down from S3
		- And ImportOperationTest for a decent sized resource and test data (patient-large.ndjson - 2 patient resources)
- [x] [bulkdata preflight checks should verify output provider config #2709](https://github.com/IBM/FHIR/issues/2709)
		- Move the Client Generation to the Server (AmazonS3)
		- Added cos to the fhir-server/pom.xml 
		- Enhanced preflight with a write check
		- Example warning that occurs with the operation now (storageProvider id/bad bucketName)
```
{
    "resourceType": "OperationOutcome",
    "id": "c0-a8-56-25-977e7573-20c9-4a0b-ba92-b07e640ddc1c",
    "issue": [
        {
            "severity": "error",
            "code": "exception",
            "diagnostics": "The bucket does not exist &#39;minio/fhirbulkdataxx&#39;"
        }
    ]
}
```
- [x] Docs: Update the Return Code from 202 to 404
- [x] Added BATCH_DB_HOSTNAME to postgres bulkdata.xml configDropin
- [x] [Bulk Data import job with pod restarts unable to delete #3033](https://github.com/IBM/FHIR/issues/3033)
		- Added a missing variable defaultHostName (we used internally however did not move to opensource)
- [x] [Datasource hits an I/IO exception with bulkdata datasource #3039](https://github.com/IBM/FHIR/issues/3039)
		- Add validationTimeout="30s" to the bulkdata datasources
- [x] [Bulk Data S3Provider with Prefix Matrix's import status does not have correct count when the job is completed #3038](https://github.com/IBM/FHIR/issues/3038)
		- Add an intermediate step to organize counts and then sort and then convert to ExitStatus
- [x] [download url is improperly constructed for ndjson exports where storageProvider is parquet-enabled
#3053](https://github.com/IBM/FHIR/issues/3053)
		- Enabled the switching between parquet and ndjson based on the output format and parquet enablement in the configuration
- [Bulk Data import fails when disableOperationOutcomes is false #2910](https://github.com/IBM/FHIR/issues/2910)
		- Enabled OperationOutcomes to be Recorded for S3
		- Added Prefixes to the Outcomes
		- Added to the BatchContextAdapter to propagate the prefixes.
		- Switched the default to disable operationoutcomes

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>
